tripal/t4d8
# Tripal content types: full support for TripalTerms

This PR adds full support for Tripal Terms to Tripal Content Types. This was started by @spficklin with #244 in the Chado Prepare and with previous implementation of TripalTerms.

#### How to Test
1. Build a new version of the Tripal Docker. This is needed to ensure the prepare step is run fresh on a clean site. Simply running prepare on an existing site may not be sufficient as the content types already exist.

```
git clone https://github.com/tripal/t4d8.git t4d8-256
cd t4d8-256
git checkout tv4g1-issue256-tripalContentType-termsAttached
docker build --tag=tripalproject/tripaldocker:t4d8-256 --build-arg drupalversion='9.4.x-dev' ./
```

The image should build without error. This shows that prepare can still be run + create content types without throwing exceptions.

2. Now create a container to do your testing in.

```
docker run --publish=9000:80 --name=t4d8-256 -tid --volume=`pwd`:/var/www/drupal9/web/modules/contrib/tripal tripalproject/tripaldocker:t4d8-256
docker exec t4d8-256 service postgresql restart
```

All further steps take place in this container through the web browser: localhost

3. Go to the Tripal Content Types listing (Admin > Tripal > Page Structure) and check that all content types are created and that the term column is filled out. This shows the prepare attached terms properly and that the list builder is able to retrieve them.
<img width="1521" alt="Screen Shot 2022-12-26 at 1 03 46 PM" src="https://user-images.githubusercontent.com/1566301/209577096-714b7bb7-a684-49a8-82e5-07aaa0581403.png">

4. Edit any content type and check that the term is fully described at the top of the page and the autocomplete is filled out and disabled. Save the content type again to make sure it doesn't change the term or throw and error on update.
<img width="1521" alt="Screen Shot 2022-12-26 at 1 04 11 PM" src="https://user-images.githubusercontent.com/1566301/209577107-0be0788e-50dc-4823-961f-645e3794079d.png">

5. Now create a new content type by clicking "Add Tripal Content Type" at the top of the Tripal Page Structure page. Fill in all the required fields. Confirm the controlled vocabulary term autocomplete pulls up values. Try the following and ensure you get validation errors:

- Change the ID Space to something that doesn't exist but still match the format (format is `NAME (ID SPACE:ACCESSION)`)
- Change the accession to something that doesn't exist but still match the format.
- Put randomness that does not match the format.
- Enter a term that is already attached to another content type (example: `gene (SO:0000704)`).

<img width="1521" alt="Screen Shot 2022-12-26 at 1 04 11 PM" src="https://user-images.githubusercontent.com/1566301/209577392-0c23d4a3-80ad-4a4e-82b3-b9dd59d46143.png">

6. Finally fill out the form in the above step correctly and click save. Check the new content type is in the list with the correct term associated with it.

<img width="1483" alt="Screen Shot 2022-12-26 at 1 12 04 PM" src="https://user-images.githubusercontent.com/1566301/209577547-1b648a2d-fba7-4dc9-bccc-80a14d1b885c.png">


#### PR Summary of Changes
✅ Pass term object to create and it saves IDSpace and Accession + sets help text based on term definition if not already set
✅ The content type ID/Name (i.e. bio_data\d+) is not automatically created when the content type is saved
✅ If not set already, the category, hide_empty_field and ajax_field are set with sensible details on save.
✅ Label, help text, term IDSpace + Accession are now checked/required on save (exception thrown).
✅ There is a new setTerm() method to allow you to set the term details with a term object rather then separately.
✅ Upgraded edit/create tripalentitytype form to show the attached term if set.
✅ Updated some tests which didn't meet the new requirements for content types.
✅ Edit/create form now has the same term autocomplete that field settings do. It is set via form state on create and via the term on update. It's disabled on update.
✅ Added to the validate of the edit/create for. It now validates the following:

- All required values were supplied (validated by Drupal automatically.
- No entity with the same label already exists
- The term was entered in the correct format: NAME (IDSPACE:ACCESSION)
- No entity exists with the same term
- The term already exists (separate message for missing ID Space vs. Accession)

✅ The term is added to the entity during the submit
✅ The content type listing is updated to display the term
✅ Fixed bug in the entity where it set the ID Space in setTermAccession() instead of the accession
✅ Fixed a bug where you couldn't save an existing content type because it tried to re-assign the id.
